src/main/drivers/motor.c

```diff
@@ -0,0 +1,251 @@
+#if defined(USE_DSHOT) && defined(USE_DSHOT_TELEMETRY)
+        if (!motorDevice->vTable.updateStart()) {
+            return;
+        }
+#endif
+        for (int i = 0; i < motorDevice->count; i++) {
+            motorDevice->vTable.write(i, values[i]);
+        }
+        motorDevice->vTable.updateComplete();
+    }
+#endif
+}
+
+int motorCount(void)
+{
+    return motorDevice->count;
+}
+
+// This is not motor generic anymore; should be moved to analog pwm module
+static void analogInitEndpoints(float outputLimit, float *outputLow, float *outputHigh, float *disarm, float *deadbandMotor3dHigh, float *deadbandMotor3dLow) {
+    if (featureIsEnabled(FEATURE_3D)) {
+        float outputLimitOffset = (flight3DConfig()->limit3d_high - flight3DConfig()->limit3d_low) * (1 - outputLimit) / 2;
+        *disarm = flight3DConfig()->neutral3d;
+        *outputLow = flight3DConfig()->limit3d_low + outputLimitOffset;
+        *outputHigh = flight3DConfig()->limit3d_high - outputLimitOffset;
+        *deadbandMotor3dHigh = flight3DConfig()->deadband3d_high;
+        *deadbandMotor3dLow = flight3DConfig()->deadband3d_low;
+    } else {
+        *disarm = motorConfig()->mincommand;
+        *outputLow = motorConfig()->minthrottle;
+        *outputHigh = motorConfig()->maxthrottle - ((motorConfig()->maxthrottle - motorConfig()->minthrottle) * (1 - outputLimit));
+    }
+}
+
+// End point initialization is called from mixerInit before motorDevInit; can't use vtable...
+void motorInitEndpoints(float outputLimit, float *outputLow, float *outputHigh, float *disarm, float *deadbandMotor3dHigh, float *deadbandMotor3dLow)
+{
+    switch (motorConfig()->dev.motorPwmProtocol) {
+#ifdef USE_DSHOT
+    case PWM_TYPE_PROSHOT1000:
+    case PWM_TYPE_DSHOT1200:
+    case PWM_TYPE_DSHOT600:
+    case PWM_TYPE_DSHOT300:
+    case PWM_TYPE_DSHOT150:
+        dshotInitEndpoints(outputLimit, outputLow, outputHigh, disarm, deadbandMotor3dHigh, deadbandMotor3dLow);
+        break;
+#endif
+    default:
+        analogInitEndpoints(outputLimit, outputLow, outputHigh, disarm, deadbandMotor3dHigh, deadbandMotor3dLow);
+        break;
+    }
+}
+
+float motorConvertFromExternal(uint16_t externalValue)
+{
+    return motorDevice->vTable.convertExternalToMotor(externalValue);
+}
+
+uint16_t motorConvertToExternal(float motorValue)
+{
+    return motorDevice->vTable.convertMotorToExternal(motorValue);
+}
+
+static bool isDshot = false; // XXX Should go somewhere else
+
+static bool motorEnableNull(void)
+{
+    return false;
+}
+
+static void motorDisableNull(void)
+{
+}
+
+bool motorUpdateStartNull(void)
+{
+    return true;
+}
+
+void motorWriteNull(uint8_t index, float value)
+{
+    UNUSED(index);
+    UNUSED(value);
+}
+
+static void motorWriteIntNull(uint8_t index, uint16_t value)
+{
+    UNUSED(index);
+    UNUSED(value);
+}
+
+void motorUpdateCompleteNull(void)
+{
+}
+
+static void motorShutdownNull(void)
+{
+}
+
+static float motorConvertFromExternalNull(uint16_t value)
```

Comment:
Where are these used?

Reply:
These are used for `motorNullDevice`, whose pointer is returned when `motorDevInit` fails. It is a protection against null function pointer call as we are not checking them for null atm.

Reply:
So is there actually a configuration that is capable of working when all motor values are constantly converted to 0? Or why are we even trying to run anything motor related in this case?
I think we really need to stop to obfuscate error conditions by pre-assigning meaningless 'default functions' that make no sense, and do not actually make anything work.

Reply:
An situation which I assume here for entirely null motor device is with the future ESCs and associated transports which are capable of detecting missing/non-responding ESCs and decides not to initialize the motor layer. User still can continue to work with rest of the system in this case.

(It will also come in handy when porting to a new MCU, while motor drivers are not finished, but this is an exceptional case, I think.)

Reply:
I don't know - to me catering to future extensions by stopping them from failing fast if they are incorrect is the wrong way to do this - if anything, an incomplete protocol implementation should be made to cause a hard fault. If the motor layer is not initialised then this should probably be handled differently than by just ignoring it - well possible that in this case it is half - initialised as well.